File: 876. Middle of the Linked List.go
// 0ms, 100.00%; 2MB, 100.00%
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
func middleNode(head *ListNode) *ListNode {
	var fast = head
	for fast.Next != nil {
		head = head.Next
		fast = fast.Next
		if fast.Next == nil {
			break
		}
		fast = fast.Next
	}
	return head
}

File: 801 - 1000/876 Middle of the Linked List.cc
// 0ms, 100.00%; 8.5MB, 88.14%
#include <bits/stdc++.h>
using namespace std;

struct ListNode {
    int val;
    ListNode *next;
    ListNode(int x) : val(x), next(NULL) {}
};

class Solution {
public:
    ListNode *middleNode(ListNode *head) {
        ListNode *fast = head;
        while (fast->next) {
            fast = fast->next;
            if (fast->next) fast = fast->next;
            head = head->next;
        }
        return head;
    }
};
